src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,382 @@
+                int entryIndex = GetEntryIndex(bucketIndex);
+
+                for(int i = entryIndex; i != -1; i = entries[i].next)
+                {
+                    if (entries[i].key.Equals(key))
+                    {
+                        return ref entries[i].value;
+                    }
+                }
+
+                if (_freeList != -1)
+                {
+                    entryIndex = _freeList;
+                    _freeList = entries[_freeList].next;
+                }
+                else
+                {
+                    if (Count == entries.Length)
+                    {
+                        entries = Resize();
+                        bucketIndex = (int)(GetHashCode(key) % (uint)entries.Length);
+                        // entry indexes were not changed by Resize
+                    }
+                    entryIndex = Count;
+                }
+
+                entries[entryIndex].key = key;
+                entries[entryIndex].next = _buckets[bucketIndex] - 1;
+                _buckets[bucketIndex] = entryIndex + 1;
+                Count++;
+                return ref entries[entryIndex].value;
+            }
+        }
+
+        private Entry[] Resize()
+        {
+            int count = Count;
+            int newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+            _entries = entries;
+
+            var newBuckets = new int[newSize];
+            while (count-- > 0)
+            {
+                int bucketIndex = (int)(GetHashCode(entries[count].key) % (uint)newBuckets.Length);
+                entries[count].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = count + 1;
+            }
+            _buckets = newBuckets;
+
+            return entries;
+        }
+
+        public IEnumerable<TKey> Keys
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                for (int i = 0; i < Count; i++)
+                {
+                    yield return entries[i].key;
+                }
+            }
+        }
+
+        public IEnumerable<TValue> Values
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                for (int i = 0; i < Count; i++)
+                {
+                    yield return entries[i].value;
+                }
+            }
+        }
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index;
+            private int _found;
+
+            private readonly HashSet<int> _freeEntries;
```

Comment:
This seems extremely expensive for enumeration. Could you instead use `~index` in the `Entry.next` field for your free list making everything less than 0 in the free list.

Reply:
Yes that was temporary :)

Reply:
`~index` is the right solution I think.